### PR TITLE
a sad tale of unicode account names, shells and code pages

### PR DESCRIPTION
**Pull request by shiftkey**
Notice anything wrong with this picture?

I'll give you a hint:

I'm not quite familiar with how Git Bash is differentiating between these two, but it's unimportant to the problem at hand, which is #2585. It took a while to get to the bottom of why the Atom installation wasn't being detected for @DRSAgile, but it's related to having non-ASCII characters in their account name.

Here's my test account running [this example script](https://gist.github.com/shiftkey/aa650e7238b41fb8c720f34fc2e74358) to read the registry:

```
C:\Users\hösentägen\Desktop\testing>node index.js
got: [
  {
    "name": "DisplayIcon",
    "type": "REG_SZ",
    "value": ""C:\\Users\\h�sent�gen\\AppData\\Local\\atom\\app.ico"
  },
  {
    "name": "DisplayName",
    "type": "REG_SZ",
    "value": "Atom"
  },
  {
    "name": "DisplayVersion",
    "type": "REG_SZ",
    "value": "1.19.5"
  },
  {
    "name": "InstallDate",
    "type": "REG_SZ",
    "value": "20170906"
  },
  {
    "name": "InstallLocation",
    "type": "REG_SZ",
    "value": ""C:\\Users\\h�sent�gen\\AppData\\Local\\atom"
  },
  {
    "name": "Publisher",
    "type": "REG_SZ",
    "value": "GitHub Inc."
  },
  {
    "name": "QuietUninstallString",
    "type": "REG_SZ",
    "value": "\""C:\\Users\\h�sent�gen\\AppData\\Local\\atom\\Update.exe\" --uninstall -s"
  },
  {
    "name": "UninstallString",
    "type": "REG_SZ",
    "value": "\""C:\\Users\\h�sent�gen\\AppData\\Local\\atom\\Update.exe\" --uninstall"
  },
  {
    "name": "EstimatedSize",
    "type": "REG_DWORD",
    "value": "0x2bc0e"
  },
  {
    "name": "NoModify",
    "type": "REG_DWORD",
    "value": "0x1"
  },
  {
    "name": "NoRepair",
    "type": "REG_DWORD",
    "value": "0x1"
  },
  {
    "name": "Language",
    "type": "REG_DWORD",
    "value": "0x409"
  }
]
```

It's those � characters that are the root cause of the issue.

If you're particularly sharp you might have noticed that directory path in the prompt is rendered correctly. Your eyes aren't deceiving you.

**But it's 2017! Doesn't Windows support Unicode?**

Yes, Windows supports Unicode character sets and code pages. But specific parts of the Windows operating system, for compatibility reasons, don't have this enabled by default - like the Windows Console. It turns out we already encountered this problem in a different context when we were retrieving the `PATH` environment variable for the current user using a PowerShell script (shout out to the Atom team for finding this earlier):

```ts
  const args = [
    '-noprofile',
    '-ExecutionPolicy',
    'RemoteSigned',
    '-command',
    // Set encoding and execute the command, capture the output, and return it
    // via .NET's console in order to have consistent UTF-8 encoding.
    // See http://stackoverflow.com/questions/22349139/utf-8-output-from-powershell
    // to address https://github.com/atom/atom/issues/5063
    `
      [Console]::OutputEncoding=[System.Text.Encoding]::UTF8
      $output=[environment]::GetEnvironmentVariable('Path', 'User')
      [Console]::WriteLine($output)
    `,
  ]

  const stdout = await spawn(powershellPath, args)
  ...
```

**Why do we need to do these shenanigans?**

Compatibility, mostly. If you're running the Command Prompt in Windows 10 chances are you have the same code page set as the IBM PC supported. That's right, [Code Page 437](https://en.wikipedia.org/wiki/Code_page_437), baby!

Before going any further, I should give some back story. **Code pages** are conceptually similar to character sets, but because of History™ there's lots of platform-specific code pages still floating around. The [documented list of these](https://msdn.microsoft.com/en-us/library/windows/desktop/dd317756(v=vs.85).aspx) is pretty entertaining read, if only to understand how messy this is.

Most code pages are a superset of ASCII, so any time you see that � or garbled text you're likely dealing with that problem. This ignores discussions around fonts, but I'm not going to deal with that here.

Anyway, Windows has an **active code page** set which affects how characters are rendered in a terminal. This also affects the behaviour of Win32 APIs - for example, `ShellExecuteA` is the ANSI API for `ShellExecute` and it's behaviour is affected by the active code page. `ShellExecuteW` is the Unicode version and is unaffected by the active code page.

Oh, and the **active code page** is set globally, so you need to be really, really careful 😱

Did you know that everything is terrible?

**What can we do here?**

The excellent people over at [`node-winreg`](https://github.com/fresc81/node-winreg) pointed out that you probably need to care about the active code page [for any UTF-8 data it reads](https://github.com/fresc81/node-winreg#processing-utf-8-data), and suggested this workaround:

> An even better approach would be to extract and store the value returned by a call to `chcp` prior setting the console to UTF-8 and resetting the codepage after your script is done.

That's the approach I've gone with here, and I'll dig more into this now that I know where I'm heading.

**But doesn't Node support Unicode? Why not just use that?**

By the time Node gets the chance to read these bytes into the stream, they're already invalid. We need to tell the shell which code page to use before spawning `reg.exe` so we get the right characters back.

 - [x] add `getActiveCodePage` method to retrieve the current code page
 - [x] add some registry tests as a safety harness
 - [x] move the registry lookup code into an external script that manages the code page

### CODE REVIEW

**Comment by joshaber** [comment]
giphy-2

File: app/src/lib/registry.ts [hunk 1]
```diff
@@ -85,3 +93,32 @@ export function readRegistryKeySafe(
...
+  const exists = await pathExists(batchFilePath)
+  if (!exists) {
+    log.error(
+      `Unable to find batch script at expected location: '${batchFilePath}'`
+    )
+    return []
+  }
+
+  const activeCodePage = await getActiveCodePage()
+  if (activeCodePage) {
```
Comment (joshaber) [inline app/src/lib/registry.ts]: This branch is empty, can we remove it or was something accidentally?

File: app/test/unit/shell-test.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+describe('shell/getActiveCodePage', () => {
+  if (process.platform === 'win32') {
+    it('can resolve the current code page on Windows', async () => {
+      const codePage = await getActiveCodePage()
+      expect(codePage).to.be.greaterThan(0)
+    })
+  }
+
+  if (process.platform !== 'win32') {
```
Comment (joshaber) [inline app/test/unit/shell-test.ts]: Nit, but could this be an `else` to the above's`if`?

File: app/src/lib/shell.ts [hunk 3]
```diff
@@ -165,3 +165,40 @@ function mergeEnvironmentVariables(env: IndexLookup) {
...
+    child.stdout.on('data', (data: Buffer) => {
+      buffers.push(data)
+    })
+
+    child.on('close', (code: number, signal) => {
+      const output = Buffer.concat(buffers).toString('utf8')
+      const result = chcpOutputRegex.exec(output)
+      if (result) {
+        const value = result[1]
+        resolve(parseInt(value, 10))
```
Comment (joshaber) [inline app/src/lib/shell.ts]: We should probably handle the case where `parseInt` returns `NaN`.

File: app/src/lib/registry.ts (hunk 1 above)
Comment (shiftkey) [reply app/src/lib/registry.ts]: Accidental. Gonna :fire: it.

**Comment by shiftkey** [comment]
🤾‍♂️

**Comment by joshaber** [comment]
giphy-2

**Pull request merged**